client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -186,8 +191,143 @@ private Checkpoint completeCheckpoint(String checkpointName) {
 
     @Override
     public void resetReaderGroup(ReaderGroupConfig config) {
-        Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, config);
-        synchronizer.updateStateUnconditionally(new ReaderGroupStateInit(config, segments, getEndSegmentsForStreams(config)));
+        while (true) {
+            val state = getState();
+            switch (state.getConfigState()) {
+                case INITIALIZING:
+                    doInit(state);
+                    break;
+                case REINITIALIZING:
+                    doReinit(state);
+                    break;
+                case READY:
+                    // This boolean will help know if the update actually succeeds or not.
+                    AtomicBoolean successfullyUpdated = new AtomicBoolean(true);
+                    long newGen = state.getGeneration() + 1;
+                    synchronizer.updateState((s, updates) -> {
+                        successfullyUpdated.set(s.equals(state));
```

Comment:
we should do a specific field comparison rather than compare the whole state object.
the reason is - state object can be modified even by other updates like checkpoints etc. 
so what we really care about is if the state.generation and state.oldconfig state.newconfig and state.status are identical

Reply:
Comparing the `ReaderGroupState`'s is done through a new method `compareState`. This method compares the generation, config, newConfig and configState